File: backend-nest-backup/src/users/users.controller.ts
import { Body, Controller, Get, Param, Post, UseGuards } from '@nestjs/common';

import { AdminAuthGuard } from '../admin-auth/admin-auth.guard';
import { Roles } from '../admin-auth/roles.decorator';
import { RolesGuard } from '../admin-auth/roles.guard';
import { GoogleLoginDto, LoginUserDto, RegisterMerchantDto, RegisterUserDto, SyncUserDto } from './users.types';
import { UsersService } from './users.service';

@Controller('users')
export class UsersController {
  constructor(private readonly usersService: UsersService) {}

  @Get()
  @UseGuards(AdminAuthGuard, RolesGuard)
  @Roles('superadmin', 'admin', 'consultor')
  list() {
    return this.usersService.list();
  }

  @Get('stats')
  @UseGuards(AdminAuthGuard, RolesGuard)
  @Roles('superadmin', 'admin', 'consultor')
  stats() {
    return this.usersService.stats();
  }

  @Get(':id')
  @UseGuards(AdminAuthGuard, RolesGuard)
  @Roles('superadmin', 'admin', 'consultor')
  findOne(@Param('id') id: string) {
    return this.usersService.findById(id);
  }

  @Post('register')
  register(@Body() dto: RegisterUserDto) {
    return this.usersService.register(dto);
  }

  @Post('register-merchant')
  registerMerchant(@Body() dto: RegisterMerchantDto) {
    return this.usersService.registerMerchant(dto);
  }

  @Post('login')
  login(@Body() dto: LoginUserDto) {
    return this.usersService.login(dto);
  }

  @Post('google-login')
  googleLogin(@Body() dto: GoogleLoginDto) {
    return this.usersService.googleLogin(dto);
  }

  @Post('sync')
  sync(@Body() dto: SyncUserDto) {
    return this.usersService.sync(dto);
  }
}

@Controller()
export class HealthController {
  @Get()
  root() {
    return {
      status: 'ok',
      service: 'unbora-backend',
      message: 'API Unbora rodando',
      endpoints: {
        health: '/health',
        recommend: 'POST /api/recomendar',
        users: '/users',
        usersStats: '/users/stats',
        carousels: '/carousels',
        notifications: '/notifications',
        locations: '/locations',
        adminLogin: 'POST /admin/auth/login',
      },
    };
  }

  @Get('health')
  health() {
    return { status: 'ok', service: 'unbora-backend' };
  }
}

File: backend/src/main/java/com/unbora/api/domain/user/UsersController.java
package com.unbora.api.domain.user;

import com.unbora.api.domain.user.dto.*;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/users")
@Tag(name = "Users", description = "Operações de usuários do app e lojistas")
public class UsersController {

    private final UsersService usersService;

    public UsersController(UsersService usersService) {
        this.usersService = usersService;
    }

    @GetMapping
    @PreAuthorize("hasAnyRole('SUPERADMIN', 'ADMIN', 'CONSULTOR')")
    @SecurityRequirement(name = "bearerAuth")
    @Operation(summary = "Listar usuários (Portal Admin)")
    public List<PublicUserDto> list() {
        return usersService.list();
    }

    @GetMapping("/stats")
    @PreAuthorize("hasAnyRole('SUPERADMIN', 'ADMIN', 'CONSULTOR')")
    @SecurityRequirement(name = "bearerAuth")
    @Operation(summary = "Estatísticas de usuários (Portal Admin)")
    public UserStatsDto stats() {
        return usersService.stats();
    }

    @GetMapping("/{id}")
    @PreAuthorize("hasAnyRole('SUPERADMIN', 'ADMIN', 'CONSULTOR')")
    @SecurityRequirement(name = "bearerAuth")
    @Operation(summary = "Buscar usuário por ID")
    public PublicUserDto findOne(@PathVariable String id) {
        return usersService.findById(id);
    }

    @PostMapping("/register")
    @Operation(summary = "Cadastrar usuário comum")
    public PublicUserDto register(@Valid @RequestBody RegisterUserDto dto) {
        return usersService.register(dto);
    }

    @PostMapping("/register-merchant")
    @Operation(summary = "Cadastrar lojista/comerciante parceiro")
    public PublicUserDto registerMerchant(@Valid @RequestBody RegisterMerchantDto dto) {
        return usersService.registerMerchant(dto);
    }

    @PostMapping("/login")
    @Operation(summary = "Login de usuário / lojista com e-mail e senha")
    public PublicUserDto login(@Valid @RequestBody LoginUserDto dto) {
        return usersService.login(dto);
    }

    @PostMapping("/google-login")
    @Operation(summary = "Login social Google")
    public PublicUserDto googleLogin(@Valid @RequestBody GoogleLoginDto dto) {
        return usersService.googleLogin(dto);
    }

    @PostMapping("/sync")
    @Operation(summary = "Sincronizar sessão do app (convidado ou logado)")
    public PublicUserDto sync(@Valid @RequestBody SyncUserDto dto) {
        return usersService.sync(dto);
    }
}
